cgrok/grok.go

```diff
@@ -35,6 +36,13 @@ type Match struct {
 	gm C.grok_match_t
 	grok *Grok
 	subject string
+
+	// Iterator state
+	gname string
+	gsubstring string
+	name *C.char
+	namelen, suboffset, sublen C.int
```

Comment:
Because these get passed to C, they were being heap allocated for every call to `Next()`. Putting them in the `Match` reduced the number of allocations to a constant number, instead of O(n) allocations.